internals-js/src/__tests__/operations.test.ts

```diff
@@ -141,3 +141,105 @@ test('fragments optimization of selection sets', () => {
     }
   `);
 });
+
+describe('selection set freezing', () => {
+  it('throws if one tries to modify a frozen selection set', () => {
+    const schema = parseSchema(`
+      type Query {
+        t: T
+      }
+
+      type T {
+        a: Int
+        b: Int
+      }
+    `);
+
+    // Note that we use parseOperation to help us build selection/selection sets because it's more readable/convenient
+    // thant to build the object "programmatically".
+    const s1 = parseOperation(schema, `{ t { a } }`).selectionSet;
+    const s2 = parseOperation(schema, `{ t { b } }`).selectionSet;
+
+    const s = new SelectionSet(schema.schemaDefinition.rootType('query')!);
+
+    // Control: check we can add to the selection set while not yet frozen
+    expect(s.isFrozen()).toBeFalsy();
+    expect(() => s.mergeIn(s1)).not.toThrow();
+
+    s.freeze();
+    expect(s.isFrozen()).toBeTruthy();
+    expect(() => s.mergeIn(s2)).toThrowError('Cannot add to frozen selection: { t { a } }');
+  });
+
+  it('it does not clone non-frozen selections when adding to another one', () => {
```

Comment:
Should you have a test that checks that `updateForAddingTo` will clone if necessary or is that covered by one of these cases?

Reply:
The test using `mergeIn` indirectly test it, as it's the cloning in this method that end up ensuring `SelectionSet::add` and `SelectionSet::mergeIn` do a clone. In fact, `updateForAddingTo` is not really meant to be use externally, it's only public because there is no simple way to define a method in `Selection` that only `SelectionSet` can use (well, I can I can make it private and cheat the type-system with `prototype`, but didn't wanted).

That said, I'll admit I didn't overdo testing here, and I could probably add at least a few test to exercise fragment selections, and why not direct test for both `add` and `updateForAddingTo`.